src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1611,16 +1679,45 @@ AddTypeConversion(Node *originalAggregate, Node *newExpression)
  * with aggregates in them, this function calls the recursive aggregate walker
  * function to create aggregates for the worker nodes. Also, the function checks
  * if we can push down the limit to worker nodes; and if we can, sets the limit
- * count and sort clause list fields in the new operator node.
+ * count and sort clause list fields in the new operator node. It provides special
+ * treatment for count distinct operator if it is used in repartition subqueries.
+ * Each column in count distinct aggregate is added to target list, and group by
+ * list of worker extended operator.
  */
 static MultiExtendedOp *
 WorkerExtendedOpNode(MultiExtendedOp *originalOpNode)
 {
 	MultiExtendedOp *workerExtendedOpNode = NULL;
+	MultiNode *parentNode = ParentNode((MultiNode *) originalOpNode);
+	MultiNode *childNode = ChildNode((MultiUnaryNode *) originalOpNode);
 	List *targetEntryList = originalOpNode->targetList;
-	List *newTargetEntryList = NIL;
 	ListCell *targetEntryCell = NULL;
+	List *newTargetEntryList = NIL;
+	List *groupClauseList = copyObject(originalOpNode->groupClauseList);
 	AttrNumber targetProjectionNumber = 1;
+	WorkerAggregateWalkerContext *walkerContext =
+		palloc0(sizeof(WorkerAggregateWalkerContext));
+	walkerContext->repartitionSubquery = false;
+	walkerContext->expressionList = NIL;
+	Index nextSortGroupRefIndex = 0;
+
+	if (CitusIsA(parentNode, MultiTable) && CitusIsA(childNode, MultiCollect))
+	{
+		walkerContext->repartitionSubquery = true;
+
+		/* find max of sort group ref index */
+		foreach(targetEntryCell, targetEntryList)
+		{
+			TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+			if (targetEntry->ressortgroupref > nextSortGroupRefIndex)
+			{
+				nextSortGroupRefIndex = targetEntry->ressortgroupref;
+			}
+		}
+	}
+
+	/* next group ref index starts from max group ref index + 1 */
+	nextSortGroupRefIndex++;
```

Comment:
`nextSortGroupRefIndex ` is only valid when it is a repartitionSubquery. So, incrementing it here (i.e., for all aggregates) might confuse the users.